File: frontend/services/courseCatalogService.ts
import api from './api';
import { Config } from '../constants/config';
import type {
  CourseCatalogImportResult,
  CourseSectionResponse,
  PageResponse,
  PaginationParams,
} from '../types';

export const COURSE_CATALOG_TERMS = [
  'Spring 2026',
  'Summer 2026',
  'Fall 2026',
  'Winter 2026',
] as const;

export type CourseCatalogSortBy =
  | 'courseCode'
  | 'courseName'
  | 'subject'
  | 'catalogNumber'
  | 'section'
  | 'classNumber'
  | 'teacher'
  | 'session'
  | 'career'
  | 'instructionMode'
  | 'meetingTimes'
  | 'weeks'
  | 'openSection'
  | 'lowCostMaterials'
  | 'noCostMaterials'
  | 'department'
  | 'credit'
  | 'semester';

export type CourseCatalogSortDirection = 'ASC' | 'DESC';

export const courseCatalogService = {
  getSections: async (
    params?: PaginationParams & {
      semester?: string;
      query?: string;
      courseCode?: string;
      subject?: string;
      sortBy?: CourseCatalogSortBy;
      sortDirection?: CourseCatalogSortDirection;
    }
  ): Promise<PageResponse<CourseSectionResponse>> => {
    const response = await api.get<PageResponse<CourseSectionResponse>>(
      '/course-catalog/sections',
      {
        params: {
          semester: params?.semester,
          q: params?.query,
          courseCode: params?.courseCode,
          subject: params?.subject,
          sortBy: params?.sortBy,
          sortDirection: params?.sortDirection,
          page: params?.page ?? 0,
          size: params?.size ?? Config.PAGE_SIZE,
        },
      }
    );
    return response.data;
  },

  getAllSectionsForCourse: async (params: {
    courseCode: string;
    semester: string;
  }): Promise<CourseSectionResponse[]> => {
    const size = 100;
    const firstPage = await courseCatalogService.getSections({
      courseCode: params.courseCode,
      semester: params.semester,
      sortBy: 'section',
      sortDirection: 'ASC',
      page: 0,
      size,
    });

    if (firstPage.totalPages <= 1) {
      return firstPage.content;
    }

    const remainingPages = await Promise.all(
      Array.from({ length: firstPage.totalPages - 1 }, (_, index) =>
        courseCatalogService.getSections({
          courseCode: params.courseCode,
          semester: params.semester,
          sortBy: 'section',
          sortDirection: 'ASC',
          page: index + 1,
          size,
        })
      )
    );

    return [...firstPage.content, ...remainingPages.flatMap((page) => page.content)];
  },

  importAllowedTerms: async (): Promise<CourseCatalogImportResult> => {
    const response = await api.post<CourseCatalogImportResult>('/faculty/course-catalog/import');
    return response.data;
  },
};

File: backend/src/main/java/com/ghost/service/CourseCatalogService.java
package com.ghost.service;

import com.ghost.dto.response.CourseSectionResponse;
import com.ghost.exception.BadRequestException;
import com.ghost.mapper.CourseSectionMapper;
import com.ghost.repository.CourseSectionRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Locale;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class CourseCatalogService {

    private static final String DEFAULT_SORT_KEY = "courseCode";
    private static final Map<String, SortSpec> SORTS = Map.ofEntries(
            Map.entry("courseCode", new SortSpec("course.courseCode", true)),
            Map.entry("classCode", new SortSpec("course.courseCode", true)),
            Map.entry("courseName", new SortSpec("course.courseName", true)),
            Map.entry("title", new SortSpec("course.courseName", true)),
            Map.entry("subject", new SortSpec("course.subject", true)),
            Map.entry("classType", new SortSpec("course.subject", true)),
            Map.entry("catalogNumber", new SortSpec("course.catalogNumber", true)),
            Map.entry("section", new SortSpec("section", true)),
            Map.entry("classNumber", new SortSpec("classNumber", true)),
            Map.entry("teacher", new SortSpec("instructor", true)),
            Map.entry("instructor", new SortSpec("instructor", true)),
            Map.entry("session", new SortSpec("session", true)),
            Map.entry("career", new SortSpec("career", true)),
            Map.entry("instructionMode", new SortSpec("instructionMode", true)),
            Map.entry("meetingTimes", new SortSpec("meetingTimes", true)),
            Map.entry("weeks", new SortSpec("numberOfWeeks", false)),
            Map.entry("numberOfWeeks", new SortSpec("numberOfWeeks", false)),
            Map.entry("openSection", new SortSpec("openSection", false)),
            Map.entry("lowCostMaterials", new SortSpec("lowCostMaterialsSection", false)),
            Map.entry("noCostMaterials", new SortSpec("noCostMaterialsSection", false)),
            Map.entry("department", new SortSpec("course.departmentName", true)),
            Map.entry("credit", new SortSpec("course.credit", true)),
            Map.entry("semester", new SortSpec("semester.name", true))
    );

    private final CourseSectionRepository courseSectionRepository;
    private final CourseSectionMapper courseSectionMapper;

    public Pageable createPageRequest(int page, int size, String sortBy, String sortDirection) {
        SortSpec sortSpec = SORTS.get(normalizeSortKey(sortBy));
        if (sortSpec == null) {
            throw new BadRequestException("Unsupported course catalog sort: " + sortBy);
        }

        Sort.Direction direction = parseDirection(sortDirection);
        Sort.Order primaryOrder = new Sort.Order(direction, sortSpec.property()).nullsLast();
        if (sortSpec.ignoreCase()) {
            primaryOrder = primaryOrder.ignoreCase();
        }

        Sort sort = Sort.by(primaryOrder)
                .and(Sort.by(Sort.Order.asc("course.courseCode").ignoreCase()))
                .and(Sort.by(Sort.Order.asc("section").ignoreCase()))
                .and(Sort.by(Sort.Order.asc("classNumber").ignoreCase()));
        return PageRequest.of(page, size, sort);
    }

    @Transactional(readOnly = true)
    public Page<CourseSectionResponse> getSections(
            String semester,
            String query,
            String subject,
            Pageable pageable
    ) {
        String normalizedSemester = normalizeOptional(semester);
        String normalizedQuery = normalizeOptional(query);
        String normalizedSubject = normalizeUpperOptional(subject);
        String squashedQuery = normalizedQuery == null ? null : normalizedQuery.replaceAll("\\s+", "");

        return courseSectionRepository
                .searchSections(normalizedSemester, normalizedQuery, squashedQuery, normalizedSubject, pageable)
                .map(courseSectionMapper::toResponse);
    }

    private String normalizeOptional(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        return value.trim();
    }

    private String normalizeUpperOptional(String value) {
        String normalized = normalizeOptional(value);
        return normalized == null ? null : normalized.toUpperCase(Locale.ROOT);
    }

    private String normalizeSortKey(String sortBy) {
        String normalized = normalizeOptional(sortBy);
        return normalized == null ? DEFAULT_SORT_KEY : normalized;
    }

    private Sort.Direction parseDirection(String sortDirection) {
        if (sortDirection == null || sortDirection.isBlank()) {
            return Sort.Direction.ASC;
        }
        try {
            return Sort.Direction.fromString(sortDirection);
        } catch (IllegalArgumentException ex) {
            throw new BadRequestException("Unsupported course catalog sort direction: " + sortDirection);
        }
    }

    private record SortSpec(String property, boolean ignoreCase) {
    }
}
